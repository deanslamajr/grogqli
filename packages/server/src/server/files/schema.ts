import shortid from 'shortid';
import editJsonFile from 'edit-json-file';
import { IntrospectionQuery } from 'graphql';
import {
  CreateTemporaryOperationRecording,
  SchemasMappingsInput,
} from '@grogqli/schema';
import fs from 'fs';

import {
  doesFileExist,
  getTemporarySchemaRecordingFilename,
  mapObjectToJsonFile,
  SCHEMA_FILE_VERSION,
  getSchemaRecordingFilePath,
  TEMP_SCHEMA_FILE_VERSION,
} from './';

type TemporarySchemaRecording = TemporarySchemaRecordingVersion1;
interface TemporarySchemaRecordingVersion1 {
  version: 1;
  hash: string;
  introspectionQuery: IntrospectionQuery;
}

type PersistTempSchemaRecording = (
  schemaIntrospectionResult: IntrospectionQuery
) => Promise<string>;

export const persistTempSchemaRecording: PersistTempSchemaRecording = async (
  schemaIntrospectionResult
) => {
  // TODO generate schema hash
  const schemaHash = shortid.generate();

  const newTempSchemaRecording: TemporarySchemaRecording = {
    version: TEMP_SCHEMA_FILE_VERSION,
    hash: schemaHash,
    introspectionQuery: schemaIntrospectionResult,
  };

  const schemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  const newTempSchemaRecordingFile = editJsonFile(schemaRecordingsPath);

  if (!doesFileExist(newTempSchemaRecordingFile)) {
    mapObjectToJsonFile(newTempSchemaRecording, newTempSchemaRecordingFile);
    newTempSchemaRecordingFile.save();
  } else {
    console.log(`Schema with hash:${schemaHash} has already been recorded!`);
  }

  return schemaHash;
};

export type SchemaRecording = SchemaRecordingVersion1;
interface SchemaRecordingVersion1 {
  version: 1;
  id: string;
  name: string;
  hash: string;
  urls: string[];
  introspectionQuery: IntrospectionQuery;
}

export const getSchemaRecordingFile = async (
  schemaId: string
): Promise<SchemaRecording> => {
  const schemaRecordingFilePath = await getSchemaRecordingFilePath(schemaId);

  return JSON.parse(
    await fs.promises.readFile(schemaRecordingFilePath, 'utf8')
  );
};

type CreateNewSchemaRecording = (params: {
  schemaUrl: string;
  schemaHash: string;
}) => Promise<string>;
const createNewSchemaRecording: CreateNewSchemaRecording = async ({
  schemaHash,
  schemaUrl,
}) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  let temporarySchemaRecordingFile: TemporarySchemaRecording;
  try {
    temporarySchemaRecordingFile = JSON.parse(
      await fs.promises.readFile(tempSchemaRecordingsPath, 'utf8')
    );
  } catch (error) {
    console.error(error);
    throw new Error(`
      Error while creating new schema recording from temporary schema recording:
      Cannot find a file for the given temporary schema recording hash:${schemaHash}
    `);
  }

  const generateUniqueSchemaIdAndCreateSchemaRecordingFile = async (): Promise<SchemaRecording> => {
    const newSchemaRecording: SchemaRecording = {
      version: SCHEMA_FILE_VERSION,
      id: shortid.generate(),
      name: 'tester schema', // TODO replace with user input
      hash: schemaHash,
      urls: [schemaUrl],
      introspectionQuery: temporarySchemaRecordingFile.introspectionQuery,
    };

    const schemaRecordingFilePath = await getSchemaRecordingFilePath(
      newSchemaRecording.id
    );

    const newSchemaRecordingFile = editJsonFile(schemaRecordingFilePath);

    if (doesFileExist(newSchemaRecordingFile)) {
      return generateUniqueSchemaIdAndCreateSchemaRecordingFile();
    } else {
      mapObjectToJsonFile(newSchemaRecording, newSchemaRecordingFile);
      newSchemaRecordingFile.save();
      return newSchemaRecording;
    }
  };

  const newSchemaRecording = await generateUniqueSchemaIdAndCreateSchemaRecordingFile();

  return newSchemaRecording.id;
};

export interface UpdatedSchemasMapping {
  schemaHash: string;
  schemaId: string;
}

type ConditionallyCreateOrUpdateSchemaRecordings = (
  schemaMappings: SchemasMappingsInput[]
) => Promise<UpdatedSchemasMapping[]>;
export const conditionallyCreateOrUpdateSchemaRecordings: ConditionallyCreateOrUpdateSchemaRecordings = async (
  schemaMappings
) => {
  return Promise.all(
    schemaMappings.map<Promise<UpdatedSchemasMapping>>(
      async ({
        opsRecordingsSchemaHash,
        targetSchemaId,
        opsRecordingsSchemaUrl,
      }) => {
        // handle NEW case
        if (targetSchemaId === 'NEW') {
          const newSchemaId = await createNewSchemaRecording({
            schemaUrl: opsRecordingsSchemaUrl,
            schemaHash: opsRecordingsSchemaHash,
          });
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: newSchemaId,
          };
        } else {
          // TODO
          // handle conditionally adding any new opsRecordingsSchemaUrl's to the mapped targetSchemaId
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: targetSchemaId,
          };
        }
      }
    )
  );
};
