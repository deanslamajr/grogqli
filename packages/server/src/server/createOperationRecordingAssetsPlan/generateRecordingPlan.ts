import shortid from 'shortid';
import {
  createRecorderApolloServer,
  OperationRecordingPlan,
  RuntimeVariables,
  RootTypeRecordingsIds,
} from './createRecorderApolloServer';

type GenerateRootTypeRecordingsIds = () => RootTypeRecordingsIds;

export const generateRootTypeRecordingsIds: GenerateRootTypeRecordingsIds = () => {
  return {
    query: shortid.generate(),
    mutation: shortid.generate(),
    subscription: shortid.generate(),
  };
};

type GenerateRecordingPlan = (params: {
  parsedOpRecording: any;
  operationSDL: string;
  schemaId: string;
  variables: any;
}) => Promise<OperationRecordingPlan>;

export const generateRecordingPlan: GenerateRecordingPlan = async ({
  parsedOpRecording,
  operationSDL,
  schemaId,
  variables,
}) => {
  // This will be mutated by apolloServer
  const recordingsPlan: OperationRecordingPlan = {
    rootTypeRecordingIds: [],
    schemaId,
    typeRecordings: {},
  };

  const unusedRootTypeRecordingsIds = generateRootTypeRecordingsIds();

  const apolloServer = await createRecorderApolloServer({
    schemaId,
  });

  await apolloServer.executeOperation({
    query: operationSDL,
    variables: {
      ...variables,
      grogqliRunTimeVariables: {
        parsedOpRecording,
        recordingsPlan,
        rootTypeRecordingsIds: unusedRootTypeRecordingsIds,
      } as RuntimeVariables,
    },
  });

  return recordingsPlan;
};
