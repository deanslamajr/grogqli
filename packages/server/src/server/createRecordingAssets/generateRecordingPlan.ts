import shortid from 'shortid';
import {
  createRecorderApolloServer,
  RecordingsPlan,
  RuntimeVariables,
  RootTypeRecordingsIds,
} from './createRecorderApolloServer';

type GenerateRootTypeRecordingsIds = () => RootTypeRecordingsIds;

export const generateRootTypeRecordingsIds: GenerateRootTypeRecordingsIds = () => {
  return {
    query: shortid.generate(),
    mutation: shortid.generate(),
    subscription: shortid.generate(),
  };
};

type GenerateRecordingPlan = (params: {
  parsedOpRecording: any;
  operationSDL: string;
  schemaId: string;
  variables: any;
}) => Promise<RecordingsPlan>;

export const generateRecordingPlan: GenerateRecordingPlan = async ({
  parsedOpRecording,
  operationSDL,
  schemaId,
  variables,
}) => {
  // This will be mutated by apolloServer
  const recordingsPlan: RecordingsPlan = {
    typeRecordings: {},
  };

  // TODO add plan data for
  // * operation
  // * workflow
  // * args
  // * etc.

  const rootTypeRecordingsIds = generateRootTypeRecordingsIds();

  const apolloServer = await createRecorderApolloServer({
    schemaId,
  });

  await apolloServer.executeOperation({
    query: operationSDL,
    variables: {
      ...variables,
      grogqliRunTimeVariables: {
        parsedOpRecording,
        recordingsPlan,
        rootTypeRecordingsIds,
      } as RuntimeVariables,
    },
  });

  return recordingsPlan;
};
