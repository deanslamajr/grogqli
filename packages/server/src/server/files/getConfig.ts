import path from 'path';
import findUp from 'find-up';
import editJsonFile from 'edit-json-file';

const CONFIG_FILE_NAME = 'grogqli.json';

export interface GrogqliConfig {
  recordingsSaveDirectory: string;
  port: number;
  shouldDogFood: boolean;
}
const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: path.join(process.cwd(), 'grogqli'),
  port: 4000,
  shouldDogFood: false,
};

type GetGrogqliConfigValue = (configKey: keyof GrogqliConfig) => string;
let getGrogqliConfigValue: GetGrogqliConfigValue | undefined;

export const getConfig = async (): Promise<GetGrogqliConfigValue> => {
  // return early if the response is cached
  if (getGrogqliConfigValue) {
    return getGrogqliConfigValue;
  }

  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(
    async (directory) => {
      const hasConfig = await findUp.exists(
        path.join(directory, CONFIG_FILE_NAME)
      );
      return hasConfig ? directory : undefined;
    },
    { type: 'directory' }
  );

  let configFromFile: editJsonFile.JsonEditor;

  if (configLocation) {
    const configFileAbsolutePath = path.join(configLocation, CONFIG_FILE_NAME);
    configFromFile = editJsonFile(configFileAbsolutePath);
  }

  getGrogqliConfigValue = (configKey) => {
    if (configKey === undefined) {
      throw new Error(
        'Invalid use of getGrogqliConfigValue: must provide a key'
      );
    }
    let configValue: any;

    // First, try to resolve the value from the given config
    if (configLocation && configFromFile) {
      configValue = configFromFile.get(configKey);

      // Make 'recordingsSaveDirectory' an abs path
      if (
        configKey === 'recordingsSaveDirectory' &&
        configValue !== undefined
      ) {
        configValue = path.join(configLocation, configValue);
      }
    }

    // If config doesn't have a value, use the default
    if (configValue === undefined) {
      configValue = DEFAULT_CONFIG[configKey];
    }

    return configValue;
  };

  return getGrogqliConfigValue;
};
