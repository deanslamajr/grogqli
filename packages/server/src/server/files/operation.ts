import path from 'path';
import shortid from 'shortid';

import {
  OPERATIONS_FILENAME,
  OPERATIONS_FOLDER_NAME,
  getRecordingsRootDir,
  getSchemasFolderPath,
} from './';

export interface OperationFile {
  id: string;
  version: number;
  recordings: {
    [opRecordingId: string]: {
      id: string;
      rootTypeRecordings: {
        [rootTypeId: string]: {
          recordingId: string;
        };
      };
    };
  };
}

type CreateNewOpFile = (args: {
  opName: string;
  schemaId: string;
}) => OperationFile;

export const createNewOpFile: CreateNewOpFile = async (args) => {
  // - generate a new unique opId
  const opId = shortid.generate();
  // - add entry to grogqli/schemas/<opPlan.schemaId>/operations.json
  const opMapFile = await loadOperationsMappingFile(args.schemaId);

  // - create grogqli/operations/<opId>.json
  //   - recordings - empty object
};

type GetOpFileFromOpName = (params: {
  opName: string;
  schemaId: string;
}) => Promise<OperationFile | null>;

export const getOpFileFromOpName: GetOpFileFromOpName = async ({
  opName,
  schemaId,
}) => {
  const opsMappingFile = await loadOperationsMappingFile(schemaId);
  const opEntry = opsMappingFile.operations[opName];

  if (opEntry === undefined) {
    return null;
  }

  const operationFile = await getOperationFile(opEntry.id);

  if (operationFile === undefined) {
    return null;
  }

  return operationFile;
};

export const getOperationFile = async (
  operationId: string
): Promise<OperationFile | null> => {
  if (!operationId) {
    return null;
  }
  let operationFile: OperationFile;
  const recordingsRootDir = await getRecordingsRootDir();
  const pathToOperationFile = path.join(
    recordingsRootDir,
    OPERATIONS_FOLDER_NAME,
    `${operationId}.json`
  );

  try {
    operationFile = require(pathToOperationFile);
  } catch (error) {
    return null;
  }

  return operationFile;
};

export interface OperationsMappingFile {
  version: number;
  operations: {
    [opName: string]: {
      name: string;
      id: string;
    };
  };
}

export const loadOperationsMappingFile = async (
  schemaId: string
): Promise<OperationsMappingFile> => {
  const schemasFolderPath = await getSchemasFolderPath();
  // TODO after a feature is added that updates these `${schemaId}.json` files at runtime,
  // reevaluate whether or not this is necessary:
  // return JSON.parse(
  //   // not using require(`${schemaId}.json`) here (unlike in the resolvers)
  //   // bc we want to bypass the auto caching feature of require
  //   fs.readFileSync(path.join(schemaRecordingsPath, `${schemaId}.json`), 'utf8')
  // );
  const pathToOperationsData = path.join(
    schemasFolderPath,
    schemaId,
    OPERATIONS_FILENAME
  );
  let operationsData: OperationsMappingFile;
  try {
    operationsData = require(pathToOperationsData);
  } catch (error) {
    // TODO handle case where file doesnt exist for the given schemaId
    throw new Error(
      `TODO handle case where a operations file doesnt exist for the given schemaId. schemaId:${schemaId}`
    );
  }
  return operationsData;
};
