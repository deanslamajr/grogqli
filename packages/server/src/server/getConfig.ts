import path from 'path';
import findUp from 'find-up';

const CONFIG_FILE_NAME = 'grogqli.json';

interface GrogqliConfig {
  recordingsSaveDirectory: string;
  recordingsFilename: string;
}

const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: '/home/dhan/codes/grogqli',
  recordingsFilename: 'groql_rec'
};
export const getConfig = async (): Promise<GrogqliConfig> => {
  console.log('process.cwd()', process.cwd());
  // TODO
  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(CONFIG_FILE_NAME);

  console.log('configLocation', configLocation)

  // TODO
  // if a config doesn't have a value defined for a particular lookup
  // use the value found in DEFAULT_CONFIG
  return DEFAULT_CONFIG;
}