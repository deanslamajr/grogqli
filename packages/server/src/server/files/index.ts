import editJsonFile from 'edit-json-file';
import fs from 'fs';
import path from 'path';
import { IntrospectionQuery } from 'graphql';

import { getConfig } from './getConfig';
import { HANDLER_SESSION_CREATED } from '../graphql/schema/resolvers/subscriptions/handlerSessionCreatedResolver';

const createDirIfDoesntExist = (dirPath: string) => {
  return fs.promises.mkdir(dirPath, { recursive: true });
};

export const doesFileExist = (file: editJsonFile.JsonEditor): boolean => {
  return Object.keys(file.read()).length > 0;
};

export const mapObjectToJsonFile = <T extends object>(
  object: T,
  file: editJsonFile.JsonEditor
): void => {
  Object.entries(object).forEach(([key, value]) => {
    file.set(key, value);
  });
};

// Directory structure
//
// /grogqli
//   /schemas
const SCHEMAS_FOLDER_NAME = 'schemas';
//     index.json
//     /<schemaId>
//       operations.json
export const OPERATIONS_NAME_TO_ID_MAPPING_FILENAME = 'operations.json';
export const OPERATIONS_NAME_TO_ID_MAPPING_VERSION = 1;
//       schema.json
const SCHEMA_FILENAME = 'schema.json';
//       types.json
const TYPES_NAME_TO_ID_MAPPING_FILENAME = 'types.json';
export const TYPES_NAME_TO_ID_MAPPING_VERSION = 1;
//   /operations
export const OPERATIONS_FOLDER_NAME = 'operations';
export const OPERATIONS_RECORDING_VERSION = 1;
//     someOperationId.json
//   /types
const TYPES_FOLDER_NAME = 'types';
//     <typeId>.json
export const TYPES_FILE_VERSION = 1;
//   /workflows
export const WORKFLOWS_FOLDER_NAME = 'workflows';
//     index.json
const WORKFLOWS_NAME_TO_ID_MAPPING_FILENAME = 'index.json';
export const WORKFLOWS_NAME_TO_ID_MAPPING_VERSION = 1;
//     <workflowId>.json
//   /local
const TEMP_FOLDER_NAME = 'local';
//     appState.json
//     /sessions
const HANDLER_SESSIONS_STATE_FOLDER_NAME = 'sessions';
//      index.json
const HANDLER_SESSION_STATE_FILENAME = 'index.json';
export const HANDLER_SESSIONS_STATE_FILE_VERSION = 1;
//      /schemas
const TEMP_SCHEMAS_FOLDER_NAME = 'schemas';
export const TEMP_SCHEMA_FILE_VERSION = 1;
//     /operations
const TEMP_OP_RECORDINGS_FOLDER_NAME = 'operations';
export const TEMP_OP_RECORDING_FILE_VERSION = 1;

// e.g. /grogqli
export const getRecordingsRootDir = async (): Promise<string> => {
  const config = await getConfig();
  const recordingsRootDir = config('recordingsSaveDirectory');
  await createDirIfDoesntExist(recordingsRootDir);
  return recordingsRootDir;
};

// e.g. /grogqli/schemas
export const getSchemasFolderPath = async (): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const schemasFolderPath = path.join(recordingsRootDir, SCHEMAS_FOLDER_NAME);
  await createDirIfDoesntExist(schemasFolderPath);
  return schemasFolderPath;
};

// e.g. /grogqli/schemas/<schemaId>/operations.json
export const getOpNameMappingFilePath = async (schemaId: string) => {
  const schemasFolderPath = await getSchemasFolderPath();

  const opNameMappingFilePath = path.join(schemasFolderPath, schemaId);

  await createDirIfDoesntExist(opNameMappingFilePath);

  return path.join(
    opNameMappingFilePath,
    OPERATIONS_NAME_TO_ID_MAPPING_FILENAME
  );
};

// e.g. /grogqli/schemas/<schemaId>/types.json
export const getTypeNameMappingFilePath = async (schemaId: string) => {
  const schemasFolderPath = await getSchemasFolderPath();

  const typeNameMappingFilePath = path.join(schemasFolderPath, schemaId);

  await createDirIfDoesntExist(typeNameMappingFilePath);

  return path.join(typeNameMappingFilePath, TYPES_NAME_TO_ID_MAPPING_FILENAME);
};

// e.g. /grogqli/workflows
export const getWorkflowsFolderPath = async (): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const workflowsFolderPath = path.join(
    recordingsRootDir,
    WORKFLOWS_FOLDER_NAME
  );
  await createDirIfDoesntExist(workflowsFolderPath);
  return workflowsFolderPath;
};

// e.g. /grogqli/workflows/index.json
export const getWorkflowNameMappingFilePath = async (): Promise<string> => {
  const workflowsFolderPath = await getWorkflowsFolderPath();
  return path.join(workflowsFolderPath, WORKFLOWS_NAME_TO_ID_MAPPING_FILENAME);
};

// e.g. /grogqli/workflows/<workflowId>.json
export const getWorkflowRecordingFilePath = async (
  workflowId: string
): Promise<string> => {
  const workflowsFolderPath = await getWorkflowsFolderPath();
  return path.join(workflowsFolderPath, `${workflowId}.json`);
};

// e.g. /grogqli/operations/<opId>.json
export const getOperationRecordingsFilePath = async (
  opId: string
): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const pathToOperationRecordingFile = path.join(
    recordingsRootDir,
    OPERATIONS_FOLDER_NAME
  );

  await createDirIfDoesntExist(pathToOperationRecordingFile);

  return path.join(pathToOperationRecordingFile, `${opId}.json`);
};

// e.g. /grogqli/types/<typeId>.json
export const getTypeFilePath = async (typeId: string) => {
  const recordingsRootDir = await getRecordingsRootDir();

  const typeFilePath = path.join(recordingsRootDir, TYPES_FOLDER_NAME);

  await createDirIfDoesntExist(typeFilePath);

  return path.join(typeFilePath, `${typeId}.json`);
};

// e.g. /grogqli/local/schemas
export const getTemporarySchemaRecordingsFolderPath = async (): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const schemaRecordingsPath = path.join(
    recordingsRootDir,
    TEMP_FOLDER_NAME,
    TEMP_SCHEMAS_FOLDER_NAME
  );
  await createDirIfDoesntExist(schemaRecordingsPath);
  return schemaRecordingsPath;
};

// e.g. /grogqli/local/schemas/<schemaHash>.json
type GetTemporarySchemaRecordingFilename = (
  schemaHash: string
) => Promise<string>;
export const getTemporarySchemaRecordingFilename: GetTemporarySchemaRecordingFilename = async (
  schemaHash
) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingsFolderPath();
  return path.join(tempSchemaRecordingsPath, `${schemaHash}.json`);
};

// e.g. /grogqli/local/sessions/<sessionId>
const getSessionFolderPath = async (sessionId: string) => {
  const recordingsRootDir = await getRecordingsRootDir();

  const givenSessionsPersistencePath = path.join(
    recordingsRootDir,
    TEMP_FOLDER_NAME,
    HANDLER_SESSIONS_STATE_FOLDER_NAME,
    sessionId
  );

  await createDirIfDoesntExist(givenSessionsPersistencePath);

  return givenSessionsPersistencePath;
};

// e.g. /grogqli/local/sessions/<sessionId>/index.json
export const getSessionsFilePath = async (sessionId: string) => {
  const givenSessionsPersistencePath = await getSessionFolderPath(sessionId);

  return path.join(
    givenSessionsPersistencePath,
    HANDLER_SESSION_STATE_FILENAME
  );
};

// e.g. /grogqli/local/sessions/<sessionId>/operations
export const getTempOpRecordingsFolderPath = async (
  sessionId: string
): Promise<string> => {
  const givenSessionsPersistencePath = await getSessionFolderPath(sessionId);

  const tempOpRecordingFolderPath = path.join(
    givenSessionsPersistencePath,
    TEMP_OP_RECORDINGS_FOLDER_NAME
  );

  await createDirIfDoesntExist(tempOpRecordingFolderPath);

  return tempOpRecordingFolderPath;
};

// e.g. /grogqli/local/sessions/<sessionId>/operations/<tempOpRecordingId>.json
type GetTempOpRecordingFilePath = (params: {
  sessionId: string;
  tempOpRecordingId: string;
}) => Promise<string>;
export const getTempOpRecordingFileName: GetTempOpRecordingFilePath = async ({
  sessionId,
  tempOpRecordingId,
}) => {
  const tempOpRecordingFolderPath = await getTempOpRecordingsFolderPath(
    sessionId
  );

  return path.join(tempOpRecordingFolderPath, `${tempOpRecordingId}.json`);
};

// Schema
// ***
// **
// *
export interface SchemaFile {
  id: string;
  version: number;
  introspectionQuery: IntrospectionQuery;
}

export const getSchema = async (schemaId: string): Promise<SchemaFile> => {
  const schemasFolderPath = await getSchemasFolderPath();
  const absPathToSchema = path.join(schemasFolderPath, schemaId);

  await createDirIfDoesntExist(absPathToSchema);

  const absPathToSchemaFile = path.join(absPathToSchema, SCHEMA_FILENAME);

  return JSON.parse(await fs.promises.readFile(absPathToSchemaFile, 'utf8'));
};
