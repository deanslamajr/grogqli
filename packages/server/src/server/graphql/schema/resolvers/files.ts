import editJsonFile from 'edit-json-file';
import fs from 'fs';

import { getConfig } from '../../../getConfig';

export const getRecordingFile = async (): Promise<editJsonFile.JsonEditor> => {
  const config = await getConfig();
  const recordingsSaveDirectory = config('recordingsSaveDirectory');
  if (!fs.existsSync(recordingsSaveDirectory)) {
    fs.mkdirSync(recordingsSaveDirectory);
  }

  return editJsonFile(
    `${recordingsSaveDirectory}/${config('recordingsFilename')}.json`
  );
};
