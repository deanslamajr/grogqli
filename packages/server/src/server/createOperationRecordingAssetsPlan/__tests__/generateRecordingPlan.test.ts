import path from 'path';

import { generateRecordingPlan } from '../generateRecordingPlan';
import { getConfig } from '../../files/getConfig';

jest.mock('../../files/getConfig');
jest.mock('shortid', () => {
  let generateCounter = 0;

  return {
    generate: () => `mockRecordingId${generateCounter++}`,
  };
});

describe('generateRecordingPlan', () => {
  const schemaId = 'anotherSchemaId';
  const parsedOpRecording = {
    data: {
      search: {
        artists: {
          nodes: [
            {
              id:
                'QXJ0aXN0OmJjNzEwYmNmLTg4MTUtNDJjZi1iYWQyLTNmMWQxMjI0NmFlYg==',
              name: 'Nickelback',
              __typename: 'Artist',
            },
            {
              id:
                'QXJ0aXN0OjFkNmI5M2JhLTg1MWUtNDQ3Yy1iOTg3LThkMzA4ZDZhYzg5ZA==',
              name: 'A Motown Tribute to Nickelback',
              __typename: 'Artist',
            },
          ],
          __typename: 'ArtistConnection',
        },
        __typename: 'SearchQuery',
      },
    },
  };
  const operationSDL =
    'query SearchForArtist {\n  search {\n    artists(query: "nickelback") {\n      nodes {\n        id\n        name\n        __typename\n      }\n      __typename\n    }\n    __typename\n  }\n}\n';
  const variables = {};

  beforeAll(() => {
    const mockedGetConfig = getConfig as jest.MockedFunction<typeof getConfig>;
    mockedGetConfig.mockImplementation(async () => () => {
      return path.join(__dirname, 'grogqli');
    });
  });

  it('should return a recording plan', async () => {
    const actual = await generateRecordingPlan({
      parsedOpRecording,
      operationSDL,
      schemaId,
      variables,
    });

    expect(actual).toMatchSnapshot();
  });
});
