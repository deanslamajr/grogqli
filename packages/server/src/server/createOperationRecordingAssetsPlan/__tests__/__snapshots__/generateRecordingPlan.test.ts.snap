// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateRecordingPlan should return a recording plan 1`] = `
Object {
  "name": "SearchForArtist",
  "rootTypeRecordingIds": Set {
    "mockRecordingId0",
  },
  "schemaId": "anotherSchemaId",
  "typeRecordings": Object {
    "mockRecordingId0": Object {
      "typeName": "Query",
      "value": Object {
        "search": "mockRecordingId3",
      },
    },
    "mockRecordingId3": Object {
      "typeName": "SearchQuery",
      "value": Object {
        "artists": "mockRecordingId4",
      },
    },
    "mockRecordingId4": Object {
      "typeName": "ArtistConnection",
      "value": Object {
        "nodes": Array [
          "mockRecordingId5",
          "mockRecordingId6",
        ],
      },
    },
    "mockRecordingId5": Object {
      "typeName": "Artist",
      "value": Object {
        "id": "QXJ0aXN0OmJjNzEwYmNmLTg4MTUtNDJjZi1iYWQyLTNmMWQxMjI0NmFlYg==",
        "name": "Nickelback",
      },
    },
    "mockRecordingId6": Object {
      "typeName": "Artist",
      "value": Object {
        "id": "QXJ0aXN0OjFkNmI5M2JhLTg1MWUtNDQ3Yy1iOTg3LThkMzA4ZDZhYzg5ZA==",
        "name": "A Motown Tribute to Nickelback",
      },
    },
  },
}
`;
